File: lzcomplexity.cpp
#include <lz/lzApp.h>

#include "main/config.h"
#include "main/main.h"

void save_data(lz::utils::LZ_Flags& flags, lz::utils::LZ_Output& results, lz_options& opt) {
   nlohmann::json out_data;

   out_data["filename"] = opt.input;
   out_data["format"]   = opt.input_format;
   out_data["size"]     = flags.input.size();

   for (std::size_t i = 0; i < flags.input.size(); i++) {
      // out_data["input"] = opt.input;
      out_data["sequences"][i]["size"] = flags.input[i].size();

      auto lz_result = results.data[i];

      out_data["sequences"][i]["lz76Complexity"]     = lz_result.getComplexity();
      out_data["sequences"][i]["lz76EntropyDensity"] = lz_result.getEntropyDensity();

      auto rsc = lz_result.getRandomShuffleComplexity();

      out_data["sequences"][i]["lz76RandomShuffleComplexity"]["value"]             = rsc.excess_value;
      out_data["sequences"][i]["lz76RandomShuffleComplexity"]["max_block_size"]    = rsc.max_block_size;
      out_data["sequences"][i]["lz76RandomShuffleComplexity"]["multi_information"] = rsc.multi_information;
      if (rsc.summands.size() > 0) {
         out_data["sequences"][i]["lz76RandomShuffleComplexity"]["summands"] = rsc.summands;
      }

      out_data["sequences"][i]["lz76Extra"]["lz_rajski_distance"]     = lz_result.getExtras().lz_rajski_distance;
      out_data["sequences"][i]["lz76Extra"]["fh_uncertainty"]         = lz_result.getExtras().fh_uncertainty;
      out_data["sequences"][i]["lz76Extra"]["lh_uncertainty"]         = lz_result.getExtras().lh_uncertainty;
      out_data["sequences"][i]["lz76Extra"]["redundancy"]             = lz_result.getExtras().redundancy;
      out_data["sequences"][i]["lz76Extra"]["lz_pearson_coefficient"] = lz_result.getExtras().lz_pearson_coefficient;

      if (opt.args.block_size >= 0) {
         auto w_rsc = lz_result.getAllRandomShuffleComplexity();

         out_data["sequences"][i]["lz76WholeRandomShuffleComplexity"]["value"]             = w_rsc.excess_value;
         out_data["sequences"][i]["lz76WholeRandomShuffleComplexity"]["max_block_size"]    = w_rsc.max_block_size;
         out_data["sequences"][i]["lz76WholeRandomShuffleComplexity"]["multi_information"] = w_rsc.multi_information;
         if (w_rsc.summands.size() > 0) {
            out_data["sequences"][i]["lz76WholeRandomShuffleComplexity"]["summands"] = w_rsc.summands;
         }
      }

      if (!opt.multiLine)
         break;
   }

   if (opt.find_distance) {
      out_data["lz76Distance"]["RandomShuffleDistance"] = results.random_shuffle_distance;
      out_data["lz76Distance"]["InformationDistance"]   = results.info_distance;
   }

   std::ofstream out(opt.output);
   if (out.is_open() && out.good()) {
      out << out_data;
      out_data.clear();
      out.close();
   }
}

void save_factors(lz::utils::LZ_Flags& flags, std::vector<lz::internal::LZ_Result> data, lz_options& opt) {
   nlohmann::json out_data;

   out_data["filename"] = opt.input;
   out_data["format"]   = opt.input_format;
   out_data["size"]     = flags.input.size();

   for (auto i = 0ul; i < data.size(); i++) {
      out_data["factors"][i] = data[i].lzf;
   }

   std::ofstream out(opt.factors_output);
   if (out.is_open() && out.good()) {
      out << out_data;
      out_data.clear();
      out.close();
   }
}

lz::lz_int process(lz_options& opt) {
   std::vector<lz::sequence> data;

   auto           g_now         = now();
   unsigned short verbose_index = 1;

   time_point_t init_time;
   if (opt.verbose) {
      std::cout << lz::GREEN_COLOR << "2. Processing the data...\n" << lz::END_COLOR;
      std::cout << lz::GREEN_COLOR << "2." << verbose_index++ << ". Reading input data from: " << lz::END_COLOR
                << opt.input << std::endl;
   }
   if (opt.preprocess) {
      multiLineToOneLine(opt.input, data, true);
   } else {
      data = read_input(opt.input, opt.multiLine, opt.input_format);
   }

   if (opt.verbose) {
      std::cout << lz::GREEN_COLOR << "Sequence to process: " << data.size() << std::endl << std::endl << lz::END_COLOR;
   }

   //? Input flags
   lz::utils::LZ_Flags test_flags(data, opt.args);
   test_flags.shuffle_init_line = opt.excess_init_line;
   test_flags.shuffle_end_line  = opt.excess_end_line;
   //? Results
   lz::utils::LZ_Output lz(data.size());

   lz::utils::EnabledMT(opt.n_jobs);

   // App functions
   if (opt.verbose) {
      std::cout << lz::GREEN_COLOR << "2." << verbose_index++ << ". Calculating lz76 factorization\n" << lz::END_COLOR;
      init_time = now();
   }
   lz::lz76Factorization(test_flags, lz);
   if (opt.verbose) {
      const auto end_time = now();
      std::cout << "Complexity: ";
      for (auto x: lz.data)
         std::cout << x.getComplexity() << " ";
      std::cout << std::endl;
      std::cout << "Finished in: " << duration(end_time - init_time) << " s" << std::endl << std::endl;
   }

   if (!opt.factors_output.empty()) {
      std::vector<lz::internal::LZ_Result> f;
      for (auto seq: test_flags.input) {
         auto flz = lz::lz76Factors(seq);
         f.push_back(flz);
         std::cout << "Factors: [ ";
         for (auto f: flz.lzf)
            std::cout << f << " ";
         std::cout << "]" << std::endl;

         std::cout << "epsilon: " << flz.epsilon << "\n";

         auto en = lz::lz76NormalError(seq);
         auto ep = lz::lz76PoisonError(seq);
         std::cout << "Errors: " << en << " " << ep << std::endl;
         // std::string txt = seq.toString();
         // for (int i = 1; i < flz.lzf.size(); i++) {
         //    std::cout << txt.substr(flz.lzf[i - 1], flz.lzf[i]) + ".";
         // }
         // std::cout << "]" << std::endl;
      }

      save_factors(test_flags, f, opt);
   }

   // App functions
   if (opt.verbose) {
      std::cout << lz::GREEN_COLOR << "2." << verbose_index++ << ". Calculating Entropy density\n" << lz::END_COLOR;
      init_time = now();
   }
   lz::lz76EntropyDensity(test_flags, lz);
   if (opt.verbose) {
      const auto end_time = now();
      std::cout << "Entropy: ";
      for (auto x: lz.data)
         std::cout << x.getEntropyDensity() << " ";
      std::cout << std::endl;
      std::cout << "Finished in: " << duration(end_time - init_time) << " s" << std::endl << std::endl;
   }

   if (opt.verbose) {
      std::cout << lz::GREEN_COLOR << "2." << verbose_index++ << ". Calculating extra measures\n" << lz::END_COLOR;
      init_time = now();
   }
   // Excess entropy by distance
   lz::lz76ExtraMeasures(test_flags, lz);
   if (opt.verbose) {
      const auto end_time = now();
      std::cout << "LZ Rajski distance: ";
      for (auto x: lz.data)
         std::cout << x.getExtras().lz_rajski_distance << " ";
      std::cout << std::endl;
      std::cout << "FH Uncertainly: ";
      for (auto x: lz.data)
         std::cout << x.getExtras().fh_uncertainty << " ";
      std::cout << std::endl;
      std::cout << "LH Uncertainly: ";
      for (auto x: lz.data)
         std::cout << x.getExtras().lh_uncertainty << " ";
      std::cout << std::endl;
      std::cout << "LZ Redundancy: ";
      for (auto x: lz.data)
         std::cout << x.getExtras().redundancy << " ";
      std::cout << std::endl;
      std::cout << "LZ Pearson coefficient: ";
      for (auto x: lz.data)
         std::cout << x.getExtras().lz_pearson_coefficient << " ";
      std::cout << std::endl;
      std::cout << "Finished in: " << duration(end_time - init_time) << " s" << std::endl << std::endl;
   }
   if (opt.verbose) {
      std::cout << lz::GREEN_COLOR << "2." << verbose_index++
                << ". Calculating random shuffle complexity using Z sequence\n"
                << lz::END_COLOR;
      init_time = now();
   }
   lz::lz76RandomShuffleComplexity(test_flags, lz);
   if (opt.verbose) {
      const auto end_time = now();
      std::cout << "Random shuffle complexity using Z sequence: ";
      for (auto x: lz.data)
         std::cout << x.getRandomShuffleComplexity().excess_value << " ";
      std::cout << std::endl;
      for (auto i = 0ul; i < lz.data.size(); i++) {
         auto x = lz.data[i];
         if (x.getRandomShuffleComplexity().summands.size() > 0) {
            std::cout << "Shuffle entropy terms of line: " << i + 1 << " [ ";
            for (auto t: x.getRandomShuffleComplexity().summands)
               std::cout << t << " ";
            std::cout << "]\n";
         }
      }
      std::cout << "Multi information: ";
      for (auto x: lz.data)
         std::cout << x.getRandomShuffleComplexity().multi_information << " ";
      std::cout << std::endl;

      std::cout << "Finished in: " << duration(end_time - init_time) << " s" << std::endl << std::endl;
   }

   if (opt.args.block_size >= 0) {
      if (opt.verbose) {
         std::cout << lz::GREEN_COLOR << "2." << verbose_index++
                   << ". Calculating random shuffle complexity using whole sequence\n"
                   << lz::END_COLOR;
         init_time = now();
      }
      // Excess entropy by shuffle
      lz::lz76WholeRandomShuffleComplexity(test_flags, lz);
      if (opt.verbose) {
         const auto end_time = now();
         std::cout << "Random shuffle complexity using whole sequence: ";
         for (auto x: lz.data)
            std::cout << x.getAllRandomShuffleComplexity().excess_value << " ";
         std::cout << std::endl;
         for (auto i = 0ul; i < lz.data.size(); i++) {
            auto x = lz.data[i];
            if (x.getAllRandomShuffleComplexity().summands.size() > 0) {
               std::cout << "Shuffle entropy terms of line: " << i + 1 << " [ ";
               for (auto t: x.getAllRandomShuffleComplexity().summands)
                  std::cout << t << " ";
               std::cout << "]\n";
            }
         }
         std::cout << "Multi information: ";
         for (auto x: lz.data)
            std::cout << x.getAllRandomShuffleComplexity().multi_information << " ";
         std::cout << std::endl;

         std::cout << "Finished in: " << duration(end_time - init_time) << " s" << std::endl << std::endl;
      }
   }

   if (opt.find_distance) {
      if (opt.verbose) {
         std::cout << lz::GREEN_COLOR << "2." << verbose_index++
                   << ". Calculating information distance between consecutive sequences\n"
                   << lz::END_COLOR;
         init_time = now();
      }
      std::vector<double> rand_dist;

      // lz::InformationDistance(test_flags, lz);
      lz::lz76InformationDistance(test_flags, lz);
      lz::lz76RandomShuffleDistance(test_flags, lz);

      // for (int i = 10; i <= 60; i += 5) {
      //    test_flags.sa_args.block_size = i;

      //    lz::RandomShuffleDistance(test_flags, lz);
      //    rand_dist.push_back(lz.random_shuffle_distance[0]);
      // }

      // for (auto x: rand_dist)
      //    std::cout << x << " ";
      // std::cout << std::endl;

      if (opt.verbose) {
         const auto end_time = now();
         std::cout << "Info distance: ";
         for (auto x: lz.info_distance)
            std::cout << x << " ";
         std::cout << std::endl;
         std::cout << "Random info distance: ";
         for (auto x: lz.random_shuffle_distance)
            std::cout << x << " ";
         std::cout << std::endl;
         std::cout << "Finished in: " << duration(end_time - init_time) << " s" << std::endl << std::endl;
      }
   }

   auto g_end = now();
   if (opt.verbose) {
      std::cout << lz::GREEN_COLOR << "2." << verbose_index++ << ". Saving results in: " << opt.output << std::endl
                << lz::END_COLOR;
      init_time = now();
   }
   save_data(test_flags, lz, opt);

   if (opt.verbose) {
      std::cout << "Total time elapsed: " << duration(g_end - g_now) << " s" << std::endl;
   }
   lz::utils::DisabledMT();
   return EXIT_SUCCESS;
}

auto main(int argc, char const* argv[]) -> int {
   cxxopts::options options("lz",
                            "LempelZiv-76 complexity utilities as a library and also a standalone software. Suited for "
                            "complexity analysis of time series.\nSend bug reports to estevez@fisica.uh.cu or "
                            "efrenaragon96@gmail.com.\n");

   // clang-format off
   options.custom_help("[OPTIONS...] input_data")
          .set_width(120)
          .set_tab_expansion(true);
   options.allow_unrecognised_options();
   // clang-format on
   auto opt_group = options.add_options("lzcomplexity");
   opt_group("a,alphabet",
             "Alphabet cardinality. If auto it tries to guess the alphabet size.",
             cxxopts::value<std::string>()->default_value("2"),
             "value");
   opt_group("d,dlz",
             "The LZ distance is calculated between consecutive lines. Only valid for multiline files (-m "
             "option).");
   opt_group(
      "e,entropy-shuffle",
      "Random shuffle complexity with whole sequence. v1: maximum value for block shuffling, f: summands output, v2: "
      "starting line for summands output, v3: ending line for summands output. All values are optionals",
      cxxopts::value<std::vector<std::string>>()->delimiter(':')->implicit_value("a"),
      "[v1]:[f]:[v2]:[v3]");
   opt_group("f,factors", "Saves the factorization.", cxxopts::value<std::string>(), "file_name");
   opt_group("F,format",
             "Input file format. TXT for raw text format. CSV the input file is a csv array. PBM, PGM and PNM is for "
             "the family of the graphic formats.",
             cxxopts::value<std::string>()->default_value("AUTO"),
             "value");
   opt_group("h,help", "Show this message.");
   opt_group("j,jobs",
             "Number of threads.",
             cxxopts::value<lz::lz_uint>()->default_value(std::to_string(std::thread::hardware_concurrency())),
             "value");
   opt_group("l,log-base",
             "The log base value. The default is the alphabet cardinality.",
             cxxopts::value<std::string>(),
             "value");
   opt_group("m,multi-line", "Treat each line in the input stream as a different sequence.");
   opt_group("n,entropy-density", "Computes only the entropy density.");
   opt_group("o,output",
             "Output filename. Default appends to the end of input file a .json extension",
             cxxopts::value<std::string>(),
             "file_name");
   opt_group("p,partitions",
             "Number of partitions used for the parallel suffix array algorithm.",
             cxxopts::value<lz::lz_int>()->default_value("20"),
             "value");
   opt_group("v,verbose", "Verbose output.", cxxopts::value<bool>()->default_value("false"));

   // opt_group("r,process", "Clear input data.");
   //    opt_group("m,max-context", "Max context for suffix comparisons (only for caps algorithm).",
   //              cxxopts::value<lz::lz_int>()->default_value("0"), "num");
   try {
      auto result = options.parse(argc, argv);

      if (result["h"].count() || result["help"].count() || result.arguments().size() == 0) {
         std::cout << lz::GREEN_COLOR << options.help() << lz::END_COLOR;
         return EXIT_SUCCESS;
      }

      auto opt = process_args(result);

      if (process(opt)) {
         return EXIT_SUCCESS;
      } else {
         return EXIT_FAILURE;
      }
   } catch (Errors er) {
      std::string msg = "Error: " + std::to_string(er.type);
      msg += "\nMessage => " + er.msg;
      std::cerr << std::endl << print_msg(lz::utils::MSG_TYPE::ERROR, msg) << std::endl;
      // std::cerr << lz::GREEN_COLOR << options.help() << lz::END_COLOR << std::endl;
      return EXIT_FAILURE;
   } catch (std::exception err) {
      std::string msg{err.what()};
      std::cerr << std::endl << print_msg(lz::utils::MSG_TYPE::ERROR, "Fatal error BOOM!!!" + msg) << std::endl;
      return EXIT_FAILURE;
   } catch (...) {
      // std::string msg{ err.what() };
      std::cerr << std::endl << print_msg(lz::utils::MSG_TYPE::ERROR, "Fatal error BOOM!!!") << std::endl;
      return EXIT_FAILURE;
   }
}

File: crates/lzcomplexity-cli/src/bin/lzcomplexity.rs
//! `lzcomplexity` — standalone LZ76 complexity engine.
//!
//! A Rust port of the C++ `standalone/lzcomplexity.cpp`. Reads a sequence file
//! (auto-detecting the format), computes LZ76 complexity, entropy density and
//! random-shuffle effective measure complexity, and writes a JSON report whose
//! schema matches the C++ tool's `save_data`.

use std::process::ExitCode;

use clap::Parser;
use lzcomplexity_cli::formats::{self, Format};
use lzcomplexity_cli::{default_output, term};
use lzcomplexity_core as core;
use serde_json::{json, Map, Value};

const VERSION: &str = env!("CARGO_PKG_VERSION");

#[derive(Parser, Debug)]
#[command(
    name = "lzcomplexity",
    disable_version_flag = true,
    about = "LempelZiv-76 complexity engine. Suited for complexity analysis of time series.\n\
             Send bug reports to estevez@fisica.uh.cu or efrenaragon96@gmail.com."
)]
struct Cli {
    /// Input sequence file.
    #[arg(value_name = "file")]
    file: Option<String>,

    /// Alphabet cardinality: `auto` guesses the alphabet size. (default auto)
    #[arg(short = 'a', long = "alphabet", value_name = "value")]
    alphabet: Option<String>,

    /// LZ distance between consecutive lines (implies multi-line).
    #[arg(short = 'd', long = "dlz", default_value_t = false)]
    dlz: bool,

    /// Random-shuffle options `v1:f:v2:v3` (block size, summand factors, line range).
    #[arg(short = 'e', long = "rsemc-opt", value_name = "opts", num_args = 0..=1, default_missing_value = "a")]
    rsemc_opt: Option<String>,

    /// Save the factorization to this file.
    #[arg(short = 'f', long = "factors", value_name = "file_name")]
    factors: Option<String>,

    /// Input file format (TXT, CSV, TSV, PBM, PGM, PNM, DNA, RNA, FASTA). Default: guess.
    #[arg(
        short = 'F',
        long = "format",
        value_name = "value",
        default_value = "AUTO"
    )]
    format: String,

    /// Number of threads. Default: hardware concurrency.
    #[arg(short = 'j', long = "jobs", value_name = "value")]
    jobs: Option<u32>,

    /// Logarithm base. Default: the alphabet cardinality.
    #[arg(short = 'l', long = "log-base", value_name = "value")]
    log_base: Option<String>,

    /// Treat each line of the input as a separate sequence.
    #[arg(short = 'm', long = "multi-line", default_value_t = false)]
    multi_line: bool,

    /// Compute only the entropy density.
    #[arg(short = 'n', long = "entropy-density", default_value_t = false)]
    entropy_density: bool,

    /// Output filename. Default: `<input>.lz76.json`.
    #[arg(short = 'o', long = "output", value_name = "file_name")]
    output: Option<String>,

    /// Number of partitions for the parallel suffix array.
    #[arg(
        short = 'p',
        long = "partitions",
        value_name = "value",
        default_value_t = 0
    )]
    partitions: i32,

    /// Verbose output.
    #[arg(short = 'v', long = "verbose", default_value_t = false)]
    verbose: bool,

    /// Print the version and exit.
    #[arg(short = 'V', long = "version", default_value_t = false)]
    version: bool,

    /// Hide warning messages.
    #[arg(short = 'w', long = "warn-out", default_value_t = false)]
    warn_out: bool,
}

struct Options {
    input: String,
    output: String,
    factors_output: Option<String>,
    format: Format,
    multi_line: bool,
    entropy_only: bool,
    find_distance: bool,
    verbose: bool,
    args: core::lz76::LzArgs,
}

fn parse_alphabet(s: &Option<String>) -> u32 {
    match s {
        None => core::NO_ALPHABET,
        Some(v) if v.eq_ignore_ascii_case("auto") || v.is_empty() => core::NO_ALPHABET,
        Some(v) => v.parse::<u32>().unwrap_or(core::NO_ALPHABET),
    }
}

/// Parse `-e v1:f:v2:v3` — v1 is the max block size (or `a` = auto = 0), `f`
/// requests the summand terms. Mirrors `parseExcessOptions`.
fn apply_rsemc(args: &mut core::lz76::LzArgs, opt: &str) {
    let parts: Vec<&str> = opt.split(':').collect();
    if parts.is_empty() || parts[0].is_empty() {
        return;
    }
    let first = parts[0];
    if first != "a" && first != "f" {
        if let Ok(v) = first.parse::<i32>() {
            args.block_size = v;
        }
    } else if first == "a" {
        args.block_size = 0;
    }
    let factor_idx = if first == "f" { 0 } else { 1 };
    if let Some(tok) = parts.get(factor_idx) {
        if *tok == "f" {
            args.get_shuffle_terms = true;
        }
    }
}

fn build_options(cli: &Cli) -> Options {
    let input = cli.file.clone().unwrap_or_default();

    // Format: explicit -F wins; otherwise resolve AUTO by extension.
    let mut format = formats::parse_format(&cli.format);
    if format == Format::Auto {
        let ext = std::path::Path::new(&input)
            .extension()
            .and_then(|e| e.to_str())
            .unwrap_or("")
            .to_ascii_lowercase();
        format = match ext.as_str() {
            "fna" | "fasta" | "gz" => Format::Fasta,
            "csv" => Format::Csv,
            "tsv" => Format::Tsv,
            _ => Format::Auto,
        };
    }

    let mut args = core::lz76::LzArgs::new();
    args.chunks = cli.partitions;
    args.alphabet = parse_alphabet(&cli.alphabet);
    args.log_base = match &cli.log_base {
        Some(v) if v.is_empty() => args.alphabet,
        Some(v) => v.parse::<u32>().unwrap_or(args.alphabet),
        None => args.alphabet,
    };
    args.block_size = -1;
    if let Some(opt) = &cli.rsemc_opt {
        apply_rsemc(&mut args, opt);
    }

    Options {
        output: cli
            .output
            .clone()
            .filter(|o| !o.is_empty())
            .unwrap_or_else(|| default_output(&input, "lz76")),
        input,
        factors_output: cli.factors.clone(),
        format,
        multi_line: cli.multi_line || cli.dlz,
        entropy_only: cli.entropy_density,
        find_distance: cli.dlz,
        verbose: cli.verbose,
        args,
    }
}

/// Standalone entropy density — the C++ batch `lz76EntropyDensity(flags, lz)`
/// computes `c * log_alphabet(n) / n`, where the log base cancels out (so the
/// `-l` flag does not affect it). `alphabet` is the explicit `-a` value or the
/// sequence's distinct-symbol count (min 2), matching the batch path.
fn standalone_entropy(seq: &core::Sequence, complexity: u32, args: &core::lz76::LzArgs) -> f64 {
    let n = seq.len() as f64;
    if n <= 1.0 {
        return 0.0;
    }
    let alphabet = if args.alphabet == core::NO_ALPHABET {
        seq.alphabet_size()
    } else {
        args.alphabet
    }
    .max(2) as f64;
    // c / (n / log_alphabet(n)) = c * ln(n) / (n * ln(alphabet))
    complexity as f64 * n.ln() / (n * alphabet.ln())
}

fn info(verbose: bool, msg: &str) {
    if verbose {
        println!("{}", term::print_msg(term::Msg::Info, msg));
    }
}

fn run(opt: &Options) -> std::io::Result<()> {
    let seqs = formats::read_input(std::path::Path::new(&opt.input), opt.multi_line, opt.format)?;
    let n_seq = seqs.len();
    info(opt.verbose, &format!("Sequences to process: {n_seq}"));

    // Which sequences appear in the report: all when multi-line, else just [0].
    let report_upto = if opt.multi_line { n_seq } else { 1.min(n_seq) };

    let mut seq_entries: Vec<Value> = Vec::with_capacity(report_upto);
    for seq in seqs.iter().take(report_upto) {
        let res = core::lz76::lz76_factors(seq, &opt.args);
        let entropy = standalone_entropy(seq, res.factorization, &opt.args);

        let mut entry = Map::new();
        entry.insert("size".into(), json!(seq.len()));
        // C++ stores the alphabet as `std::vector<char>` (signed on x86-64), so
        // nlohmann serialises bytes >= 128 as negative numbers — match that.
        entry.insert(
            "alphabet".into(),
            Value::Array(seq.alphabet().iter().map(|&b| json!(b as i8)).collect()),
        );
        entry.insert("alphabet_size".into(), json!(seq.alphabet_size()));
        entry.insert("lz76Complexity".into(), json!(res.factorization));
        entry.insert("lz76EntropyDensity".into(), json!(entropy));

        // The C++ `save_data` always writes the random-shuffle object. In
        // entropy-only (-n) mode the shuffle stage is skipped, so it emits the
        // default-constructed values (-1 / 0.0 / 0.0, summands omitted).
        let mut rsc_map = Map::new();
        if opt.entropy_only {
            rsc_map.insert("value".into(), json!(0.0));
            rsc_map.insert("max_block_size".into(), json!(-1));
            rsc_map.insert("multi_information".into(), json!(0.0));
        } else {
            let rsc = core::shuffle::lz76_random_shuffle_complexity(seq, &opt.args);
            rsc_map.insert("value".into(), json!(rsc.emc_value));
            rsc_map.insert("max_block_size".into(), json!(rsc.max_block_size));
            rsc_map.insert("multi_information".into(), json!(rsc.multi_information));
            if !rsc.summands.is_empty() {
                rsc_map.insert("summands".into(), json!(rsc.summands));
            }
        }
        entry.insert("lz76RandomShuffleComplexity".into(), Value::Object(rsc_map));
        seq_entries.push(Value::Object(entry));
    }

    let mut out = Map::new();
    out.insert("filename".into(), json!(opt.input));
    out.insert("format".into(), json!(opt.format.magic_name()));
    out.insert("size".into(), json!(n_seq));
    out.insert("sequences".into(), Value::Array(seq_entries));

    if opt.find_distance {
        // Consecutive-sequence information distance (fmax formula). For a single
        // sequence the C++ batch returns `[0]`; otherwise n-1 pairwise values.
        let dists: Vec<f64> = if n_seq <= 1 {
            vec![0.0]
        } else {
            (0..n_seq - 1)
                .map(|i| {
                    core::metrics::lz76_information_distance(&seqs[i], &seqs[i + 1], &opt.args)
                })
                .collect()
        };
        out.insert(
            "lz76Distance".into(),
            json!({ "InformationDistance": dists }),
        );
    }

    std::fs::write(&opt.output, serde_json::to_string(&Value::Object(out))?)?;

    // Separate factors file.
    if let Some(fpath) = &opt.factors_output {
        let mut factors: Vec<Value> = Vec::new();
        for seq in seqs.iter() {
            let res = core::lz76::lz76_factors(seq, &opt.args);
            factors.push(json!(res.lzf));
        }
        // The C++ `save_factors` writes `format` as the raw enum int (unlike
        // `save_data`, which uses the MagicValues string).
        let fjson = json!({
            "filename": opt.input,
            "format": opt.format.magic_ordinal(),
            "size": n_seq,
            "factors": factors,
        });
        std::fs::write(fpath, serde_json::to_string(&fjson)?)?;
    }

    info(opt.verbose, &format!("Saved results in: {}", opt.output));
    Ok(())
}

fn main() -> ExitCode {
    let cli = Cli::parse();

    if cli.version {
        println!(
            "{}",
            term::print_msg(term::Msg::Info, &format!("v{VERSION}"))
        );
        return ExitCode::SUCCESS;
    }

    if cli.file.is_none() {
        // No positional file: with no arguments at all, print help and exit
        // success (matching the C++ tool); otherwise it is a usage error.
        if std::env::args().len() <= 1 {
            let _ = <Cli as clap::CommandFactory>::command().print_help();
            return ExitCode::SUCCESS;
        }
        eprintln!(
            "{}",
            term::print_msg(term::Msg::Error, "Input file is missing")
        );
        return ExitCode::FAILURE;
    }

    // Honour -j/--jobs by sizing the global rayon pool (mirrors EnabledMT).
    if let Some(j) = cli.jobs {
        if j > 0 {
            let _ = rayon::ThreadPoolBuilder::new()
                .num_threads(j as usize)
                .build_global();
        }
    }

    let opt = build_options(&cli);

    if !std::path::Path::new(&opt.input).exists() {
        eprintln!(
            "{}",
            term::print_msg(
                term::Msg::Error,
                &format!("File doesn't exist: {}", opt.input)
            )
        );
        return ExitCode::FAILURE;
    }

    match run(&opt) {
        Ok(()) => ExitCode::SUCCESS,
        Err(e) => {
            eprintln!("{}", term::print_msg(term::Msg::Error, &e.to_string()));
            ExitCode::FAILURE
        }
    }
}
